numpy/core/numeric.py

```diff
@@ -374,13 +378,85 @@ def extend_all(module):
         if a not in adict:
             __all__.append(a)
 
+
+def count_nonzero(a, axis=None):
+    """
+    Counts the number of non-zero values in the array ``a``.
+
+    Parameters
+    ----------
+    a : array_like
+        The array for which to count non-zeros.
+
+    axis : int, optional
+        Axis along which to count non-zeros. Default is None,
+        meaning that non-zeros will be counted along a flattened
+        version of ``a``.
+
+        .. versionadded:: 1.12.0
+
+    Returns
+    -------
+    count : int or array of int
+        Number of non-zero values in the array along a given axis.
+        Otherwise, the total number of non-zero values in the array
+        is returned.
+
+    See Also
+    --------
+    nonzero : Return the coordinates of all the non-zero values.
+
+    Examples
+    --------
+    >>> np.count_nonzero(np.eye(4))
+    4
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]])
+    5
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]], axis=0)
+    array([1, 1, 1, 1, 1])
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]], axis=1)
+    array([2, 3])
+
+    """
+    a = asanyarray(a)
+
+    if a.dtype.name == 'bool':
+        if axis is None:
+            # multiarray's count_nonzero
+            # has an optimized version for
+            # booleans on a flat array
+            return multiarray.count_nonzero(a)
+
+        return a.sum(axis=axis)
+
+    nonzeros = (a != 0)
+
+    # FutureWarning is generated if element-wise
+    # comparisons fail --> a scalar is returned
+    #
+    # In the future, the behaviour of such comparisons
+    # will change (e.g. raise an Exception), so the
+    # behaviour of this block will have to respond in kind
+    if not isinstance(nonzeros, ndarray):
+        # In these cases, the concept of 'axis' is not
+        # well understood since the objects being passed
+        # could have very unusual 'shapes' and 'dimensions'
+        if axis is not None:
+            warnings.warn(("'axis' is not well understood "
+                           "with complex-dtype arrays. Computing "
+                           "over the flattened array instead."),
+                          RuntimeWarning)
+
+        return multiarray.count_nonzero(a)
```

Comment:
Your elements are fine, since they all evaluate to True if you put them into `bool(element)`. The problem is things like an array containing arrays, for which `bool(element)` should raise an error, the error is currently not handled and evaluated as True. I am not sure what should happen, but I would almost say that we should raise the error.